From 33d549538bdba7acf91c7a01626335d6d4afe727 Mon Sep 17 00:00:00 2001
From: fundamental <mark.d.mccurry@gmail.com>
Date: Thu, 18 Jun 2015 10:42:39 -0400
Subject: [PATCH 0502/1936] UI: Capture Messages During Initialization

Previously during the construction of the GUI a number of messages
would receive responses, however these would not be routed to widgets
as the gui callback was not set until after initialization.
These messages are now captured in a temporary vector and played back
to the UI after setting the callback.
This should fix some startup weirdness which has been visible
in the past.
---
 src/main.cpp | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/main.cpp b/src/main.cpp
index b5434d08..7069506d 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -429,11 +429,29 @@ int main(int argc, char *argv[])
 
 
     gui = NULL;
+
+    //Capture Startup Responses
+    typedef std::vector<const char *> wait_t;
+    wait_t msg_waitlist;
+    middleware->setUiCallback([](void*v,const char*msg) {
+            wait_t &wait = *(wait_t*)v;
+            size_t len = rtosc_message_length(msg, -1);
+            char *copy = new char[len];
+            memcpy(copy, msg, len);
+            wait.push_back(copy);
+            }, &msg_waitlist);
+
     if(!noui)
         gui = GUI::createUi(middleware->spawnUiApi(), &Pexitprogram);
     middleware->setUiCallback(GUI::raiseUi, gui);
     middleware->setIdleCallback([](void*){GUI::tickUi(gui);}, NULL);
 
+    //Replay Startup Responses
+    for(auto msg:msg_waitlist) {
+        GUI::raiseUi(gui, msg);
+        delete [] msg;
+    }
+
     if(!noui)
     {
         GUI::raiseUi(gui, "/show",  "i", config.cfg.UserInterfaceMode);
-- 
2.47.0

